README
mailfront
Mail server network protocol front-ends
Bruce Guenter <bruce@untroubled.org>
Version 2.01
2013-07-15

This is mailfront, a package containing customizeable network front-ends
for mail servers.  It contains complete SMTP, QMQP, QMTP, and POP3
front-ends as well as an authentication module for IMAP.  The mail
delivery front-ends also contain internal address filtering features.

Two SMTP back-ends are provided.  One delivers mail to qmail-queue,
mimicking most of the behavior of qmail-smtpd, with the addition of
support for SMTP AUTH.  The other rejects all SMTP commands if
$SMTPREJECT is set, and execs its command line otherwise (in order to
run the above program).

A mailing list has been set up to discuss this and other packages.
To subscribe, send an email to:
	bgware-subscribe@lists.untroubled.org
A mailing list archive is available at:
	http://lists.untroubled.org/?list=bgware

Development versions of mailfront are available at:
	https://github.com/bruceg/mailfront

Requirements:

- bglibs version 1.101
- cvm version 0.82
- Lua version 5 or later (optional)

Installation:

- Build the sources by running "make".
- To build the Lua plugin, run "make lua".
- After the package has been compiled, run "make install" as root.

Configuration:

- To take advantage of the SMTP AUTH features, make sure you have a CVM
  authentication program (some are included with the cvm package itself).
- Run a CVM authentication module to provide the AUTH feature.
  Example:  To run cvm-vmailmgr as a daemon:
    exec /usr/local/bin/softlimit -m 9000000 \
     /usr/local/bin/cvm-vmailmgr /tmp/.cvm-vmailmgr 2>&1
- Configure your mail system to use the SMTP back-end with the
  appropriate environment variables.

  Example using tcpserver (highly recommended):
    #!/bin/sh
    QMAILDUID=`id -u qmaild`
    NOFILESGID=`id -g qmaild`
    MAXSMTPD=`head -1 /var/qmail/control/concurrencyincoming`
    if [ -z "$QMAILDUID" -o -z "$NOFILESGID" -o -z "$MAXSMTPD" ]; then
        echo $0: QMAILDUID, NOFILESGID, or MAXSMTPD is unset
        exit 1
    fi
    exec \
    /usr/local/bin/envdir /etc/smtpfront \
    /usr/local/bin/softlimit -m 2000000 \
    /usr/local/bin/tcpserver -v -R -H \
      -l "`head -1 /var/qmail/control/me`" -x /etc/tcp.smtp.cdb \
      -c "$MAXSMTPD" -u "$QMAILDUID" -g "$NOFILESGID" 0 25 \
    /usr/local/bin/smtpfront-qmail 2>&1

  /etc/smtpfront/CVM_SASL_PLAIN:
    cvm-local:/tmp/.cvm-vmailmgr

  Example using xinetd with TCP Wrappers:

  /etc/xinetd.d/smtp:
    # default: on
    # description: smtp
    service smtp
    {
    	disable	= no
    	flags		= REUSE NAMEINARGS
    	socket_type	= stream        
    	protocol	= tcp
    	wait		= no
    	user		= qmaild
    	server		= /usr/sbin/tcpd
    	server_args 	= /var/qmail/bin/tcp-env -R /usr/local/sbin/smtpfront-wrapper
    	log_on_success          += USERID
    	log_on_failure          += USERID
    }

  /usr/local/sbin/smtpfront-wrapper:
    #!/bin/sh
    CVM_SASL_PLAIN=cvm-local:/tmp/.cvm-unix
    export CVM_SASL_PLAIN
    CVM_SASL_LOGIN=cvm-local:/tmp/.cvm-unix
    export CVM_SASL_LOGIN
    exec /usr/local/bin/smtpfront-qmail 2>> /tmp/smtpfront-errs.txt


This project was initiated at FutureQuest, Inc.  We are releasing it
as an open-source project because we felt it would be useful to others,
as well as to repay our debt of gratitude to the larger open-source
community for the excellent packages we have enjoyed.

For more details, you may contact FutureQuest, Inc. at:
FutureQuest, Inc.
PO BOX 623127
Oviedo FL 32762-3127 USA
http://www.FutureQuest.net/
ossi@FutureQuest.net

This package is Copyright(C) 2013 Bruce Guenter or FutureQuest, Inc.,
and may be copied according to the GNU GENERAL PUBLIC LICENSE (GPL)
Version 2 or a later version.  A copy of this license is included with
this package.  This package comes with no warranty of any kind.


ChangeLog
------------------------------------------------------------------------
r554 | bruce | 2007-09-27 15:23:24 -0600 (Thu, 27 Sep 2007) | 2 lines
Changed paths:
   M /trunk/NEWS

Added note about API documentation.

------------------------------------------------------------------------
r553 | bruce | 2007-09-27 15:20:52 -0600 (Thu, 27 Sep 2007) | 2 lines
Changed paths:
   A /trunk/EXTRADIST
   M /trunk/makedist.py
   M /trunk/plugin-api.html
   D /trunk/plugin-prototype.c
   D /trunk/plugin-prototype.html
   A /trunk/plugin-template.c (from /trunk/plugin-prototype.c:542)
   A /trunk/plugin-template.html (from /trunk/plugin-prototype.html:542)

Small reorganization of the template plugin and links.

------------------------------------------------------------------------
r552 | bruce | 2007-09-27 14:19:06 -0600 (Thu, 27 Sep 2007) | 2 lines
Changed paths:
   M /trunk/TODO
   M /trunk/mailfront.c
   M /trunk/plugin-api.html

Made mailfront call the reset hooks early on message permanent errors.

------------------------------------------------------------------------
r551 | bruce | 2007-09-27 09:47:22 -0600 (Thu, 27 Sep 2007) | 2 lines
Changed paths:
   M /trunk/TODO
   A /trunk/tests/plugin-qmail-validate

Added tests for plugin-qmail-validate.

------------------------------------------------------------------------
r550 | bruce | 2007-09-27 09:16:21 -0600 (Thu, 27 Sep 2007) | 2 lines
Changed paths:
   M /trunk/tests.inc

Tweaked the tests to report unexpected startup messages.

------------------------------------------------------------------------
r549 | bruce | 2007-09-27 00:08:52 -0600 (Thu, 27 Sep 2007) | 2 lines
Changed paths:
   M /trunk/tests/smtpfront-auth-login
   M /trunk/tests/smtpfront-auth-plain
   M /trunk/tests/smtpfront-require-auth

Removed dependancy on qmail-validate in tests.

------------------------------------------------------------------------
r548 | bruce | 2007-09-27 00:01:29 -0600 (Thu, 27 Sep 2007) | 3 lines
Changed paths:
   M /trunk/backend-echo.c

Restored the ability of backend-echo.c to report data bytes when a
temporary file is in use.

------------------------------------------------------------------------
r547 | bruce | 2007-09-26 15:47:16 -0600 (Wed, 26 Sep 2007) | 2 lines
Changed paths:
   M /trunk/plugin-api.html

Missed a paragraph in the API document.

------------------------------------------------------------------------
r546 | bruce | 2007-09-26 14:52:29 -0600 (Wed, 26 Sep 2007) | 3 lines
Changed paths:
   M /trunk/mailfront.html
   A /trunk/plugin-api.html

Wrote a plugin API document, moving some technical information out of
the main page.

------------------------------------------------------------------------
r545 | bruce | 2007-09-25 15:38:33 -0600 (Tue, 25 Sep 2007) | 4 lines
Changed paths:
   M /trunk/TODO
   M /trunk/backend-echo.c
   M /trunk/backend-qmail.c
   M /trunk/builtins.c
   A /trunk/mailfront-internal.h (from /trunk/mailfront.h:544)
   M /trunk/mailfront.c
   M /trunk/mailfront.h
   M /trunk/modules.c
   M /trunk/plugin-add-received.c
   M /trunk/plugin-mailrules.c
   M /trunk/protocol-smtp.c
   M /trunk/session.c

Moved all internal implementation declarations into their own header,
and purged all modules of references to those internals (most notably
direct access to the session structure).

------------------------------------------------------------------------
r544 | bruce | 2007-09-25 14:51:58 -0600 (Tue, 25 Sep 2007) | 4 lines
Changed paths:
   M /trunk/mailfront.c
   M /trunk/mailfront.h
   M /trunk/modules.c

Moved several global variables into the session structure.
They have been added to the end of the structure to retain ABI
compatibility with existing plugins.

------------------------------------------------------------------------
r543 | bruce | 2007-09-25 12:36:30 -0600 (Tue, 25 Sep 2007) | 3 lines
Changed paths:
   M /trunk/plugin-add-received.html
   M /trunk/protocol-smtp.html

Document use of ${PROTO}LOCALHOST (and others) in the SMTP protocol and
the add-received plugin.

------------------------------------------------------------------------
r542 | bruce | 2007-09-24 17:20:01 -0600 (Mon, 24 Sep 2007) | 2 lines
Changed paths:
   M /trunk/makedist.py
   A /trunk/plugin-prototype.c (from /trunk/plugin.c:534)
   A /trunk/plugin-prototype.html (from /trunk/plugin.html:534)
   D /trunk/plugin.c
   D /trunk/plugin.html

Renamed the prototype to plugin-prototype.c and added it to the web site files.

------------------------------------------------------------------------
r541 | bruce | 2007-09-03 12:02:39 -0600 (Mon, 03 Sep 2007) | 2 lines
Changed paths:
   M /trunk/NEWS
   M /trunk/plugin-clamav.c
   M /trunk/plugin-clamav.html

Modified the ClamAV plugin to prefer $CLAMAV_* settings over $CLAMD_*

------------------------------------------------------------------------
r540 | bruce | 2007-01-22 15:38:27 -0600 (Mon, 22 Jan 2007) | 2 lines
Changed paths:
   M /trunk/NEWS
   M /trunk/plugin-clamav.c
   M /trunk/plugin-clamav.html

Added a maximum message size scanning limit to the clamav plugin.

------------------------------------------------------------------------
r539 | bruce | 2007-01-22 15:08:25 -0600 (Mon, 22 Jan 2007) | 2 lines
Changed paths:
   M /trunk/NEWS
   M /trunk/plugin-clamav.c
   M /trunk/plugin-clamav.html

Added a separate send timeout to the virus scanner.

------------------------------------------------------------------------
r538 | bruce | 2007-01-22 10:18:01 -0600 (Mon, 22 Jan 2007) | 3 lines
Changed paths:
   M /trunk/NEWS
   M /trunk/plugin-clamav.c

Fixed a bug in the ClamAV plugin with handling port numbers when using
multiple IPs.

------------------------------------------------------------------------
r537 | bruce | 2007-01-22 10:13:00 -0600 (Mon, 22 Jan 2007) | 2 lines
Changed paths:
   M /trunk/NEWS
   M /trunk/plugin-clamav.c
   M /trunk/plugin-clamav.html

Added a separate connect timeout to the ClamAV plugin.


imapfront-auth.c
/* imapfront-auth.c - IMAP authentication front-end
 * Copyright (C) 2008  Bruce Guenter <bruce@untroubled.org> or FutureQuest, Inc.
 * Development of this program was sponsored by FutureQuest, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Contact information:
 * FutureQuest Inc.
 * PO BOX 623127
 * Oviedo FL 32762-3127 USA
 * http://www.FutureQuest.net/
 * ossi@FutureQuest.net
 */
#include <sysdeps.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <cvm/sasl.h>
#include <cvm/v2client.h>
#include <iobuf/iobuf.h>
#include <str/str.h>
#include "ucspitls.h"

const char program[] = "imapfront-auth";
const int msg_show_pid = 1;

#define MAX_ARGC 16
#define QUOTE '"'
#define ESCAPE '\\'
#define LBRACE '{'
#define RBRACE '}'

static const char NOTAG[] = "*";
static const char CONT[] = "+";

static const char* capability;
static const char* cvm;
static const char* domain;
static char** nextcmd;
static str tag;
static str line;
static str cmd;
static str line_args[MAX_ARGC];
static int line_argc;

static int auth_count;
static int auth_max;

static struct sasl_auth saslauth = { .prefix = "+ " };

static int tls_available = 0;
static int auth_available = 1;

void log_start(const char* tagstr)
{
  obuf_puts(&errbuf, program);
  obuf_putc(&errbuf, '[');
  obuf_putu(&errbuf, getpid());
  obuf_puts(&errbuf, "]: ");
  if (tagstr) {
    obuf_puts(&errbuf, tagstr);
    obuf_putc(&errbuf, ' ');
  }
}

void log_str(const char* msg) { obuf_puts(&errbuf, msg); }
void log_end(void) { obuf_putsflush(&errbuf, "\n"); }

void logmsg(const char* tagstr, const char* msg)
{
  log_start(tagstr);
  log_str(msg);
  log_end();
}

void respond_start(const char* tagstr)
{
  if (tagstr == 0) tagstr = tag.s;
  log_start(tagstr);
  if (!obuf_puts(&outbuf, tagstr) ||
      !obuf_putc(&outbuf, ' '))
    exit(1);
}

void respond_str(const char* msg)
{
  log_str(msg);
  if (!obuf_puts(&outbuf, msg))
    exit(1);
}

void respond_end(void)
{
  log_end();
  if (!obuf_putsflush(&outbuf, CRLF))
    exit(1);
}

void respond(const char* tagstr, const char* msg)
{
  respond_start(tagstr);
  respond_str(msg);
  respond_end();
}

#if 0
static int isctl(char ch) { return (ch > 0x1f) && (ch < 0x7f); }
static int isquotedspecial(char ch) { return (ch == QUOTE) || (ch == ESCAPE); }

static int isatomspecial(char ch)
{
  switch (ch) {
  case '(':
  case ')':
  case '{':
  case ' ':
  case '%':
  case '*':
    return 0;
  }
  return !isctl(ch) && !isquotedspecial(ch);
}

static int isatom(char ch) { return !isatomspecial(ch); }
static int istag(char ch) { return isatom(ch) && (ch != '+'); }
#endif

/* This parser is rather liberal in what it accepts:
   The IMAP standard mandates seperate character sets for tags, commands,
   unquoted strings.  Since they all must be seperated by spaces, this
   parser allows all non-whitespace characters for each of the above. */
static int parse_line(void)
{
#define RESPOND(T,S) do{ respond(T,S); return 0; }while(0)
  unsigned i;
  unsigned len;
  const char* ptr;
  str* arg;
  
  /* Parse out the command tag */
  str_truncate(&tag, 0);
  for (i = 0, ptr = line.s; i < line.len && isspace(*ptr); ++i, ++ptr) ;
  for (; i < line.len && !isspace(*ptr); ++i, ++ptr)
    str_catc(&tag, line.s[i]);
  if (!tag.len) RESPOND(NOTAG, "BAD Syntax error");
  if (i >= line.len) RESPOND(0, "BAD Syntax error");

  /* Parse out the command itself */
  str_truncate(&cmd, 0);
  for (; i < line.len && isspace(*ptr); ++i, ++ptr) ;
  for (; i < line.len && !isspace(*ptr); ++i, ++ptr)
    str_catc(&cmd, line.s[i]);
  if (!cmd.len) RESPOND(0, "BAD Syntax error");

  /* Parse out the command-line args */
  for (line_argc = 0; line_argc < MAX_ARGC; ++line_argc) {
    arg = &line_args[line_argc];
    str_truncate(arg, 0);
    for (; i < line.len && isspace(*ptr); ++i, ++ptr) ;
    if (i >= line.len) break;

    switch (*ptr) {
    case LBRACE:
      /* Handle a string literal */
      ++i, ++ptr;
      if (!isdigit(*ptr)) RESPOND(0, "BAD Syntax error: missing integer");
      for (len = 0; i < line.len && *ptr != RBRACE; ++i, ++ptr) {
	if (!isdigit(*ptr))
	  RESPOND(0, "BAD Syntax error: invalid integer");
	len = len * 10 + *ptr - '0';
      }
      ++i, ++ptr;
      if (*ptr != 0) RESPOND(0, "BAD Syntax error: missing LF after integer");
      str_ready(arg, len);
      respond(CONT, "OK");
      if (len > 0)
	ibuf_read(&inbuf, arg->s, len);
      arg->s[arg->len = len] = 0;
      ibuf_getstr_crlf(&inbuf, &line);
      i = 0;
      ptr = line.s;
      break;

    case QUOTE:
      /* Handle a quoted string */
      for (++i, ++ptr; i < line.len && *ptr != QUOTE; ++i, ++ptr) {
	if (*ptr == ESCAPE) {
	  if (++i >= line.len) break;
	  ++ptr;
	}
	str_catc(arg, *ptr);
      }
      if (i >= line.len || *ptr != QUOTE)
	RESPOND(0, "BAD Syntax error: unterminated quoted string");
      ++i, ++ptr;
      break;

    default:
      /* Normal case is very simple */
      for (; i < line.len && !isspace(*ptr); ++i, ++ptr)
	str_catc(arg, *ptr);
    }
  }
  for (; i < line.len && isspace(*ptr); ++i, ++ptr) ;
  if (i < line.len) RESPOND(0, "BAD Too many command arguments");
  return 1;
}

void cmd_noop(void)
{
  respond(0, "OK NOOP completed");
}

void cmd_logout(void)
{
  respond(NOTAG, "Logging out");
  respond(0, "OK LOGOUT completed");
  exit(0);
}

void cmd_capability(void)
{
  const char *p;

  respond_start(NOTAG);
  respond_str("CAPABILITY IMAP4rev1");
  if (tls_available)
    respond_str(" STARTTLS");
  if (!auth_available)
    respond_str(" LOGINDISABLED");
  if (*capability != 0) {
    respond_str(" ");
    respond_str(capability);
  }
  if ((p = getenv("IMAP_ACL")) && atoi(p))
    respond_str(" ACL ACL2=UNION");
  if ((p = getenv("OUTBOX")) && *p) {
    respond_str(" XCOURIEROUTBOX=INBOX");
    respond_str(p);
  }
  if ((p = getenv("IMAP_MOVE_EXPUNGE_TO_TRASH")) && atoi(p))
    respond_str(" XMAGICTRASH");
  respond_end();
  respond(0, "OK CAPABILITY completed");
}

static int setup_env(void)
{
  const char* s;
  const char* colon;
  struct stat st;

  if (cvm_fact_mailbox != 0
      && (s = getenv("SETUP_ENV")) != 0
      && strcmp(s, "dovecot") == 0) {
    /* This tells Dovecot that its environment has already been set up. */
    if (putenv("DOVECONF_ENV=1") != 0)
      return 0;
    /* Use the file type to set the prefix to mbox: or maildir:
     * Assume missing files are mboxes. */
    s = (stat(cvm_fact_mailbox, &st) == 0
	 && S_ISDIR(st.st_mode))
      ? "maildir:"
      : "mbox:";
    /* Use cmd for temporary storage of the substituted mailbox name */
    if (!str_copys(&cmd, s))
      return 0;
    s = cvm_fact_mailbox;
    while ((colon = strchr(s, ':')) != 0) {
      if (!str_catb(&cmd, s, colon-s)
	  || !str_catb(&cmd, "::", 2))
	return 0;
      s = colon + 1;
    }
    if (!str_cats(&cmd, s))
      return 0;
    cvm_fact_mailbox = cmd.s;
  }
  return cvm_setenv()
    && setenv("IMAPLOGINTAG", tag.s, 1) == 0
    && setenv("AUTHENTICATED", cvm_fact_username, 1) == 0;
}

void do_exec(void)
{
  if (!cvm_setugid())
    respond(0, "NO Internal error: could not set UID/GID");
  else if (!setup_env())
    respond(0, "NO Internal error: could not set environment");
  else {
    alarm(0);
    execvp(nextcmd[0], nextcmd);
    respond(0, "NO Could not execute second stage");
  }
  exit(1);
}

void cmd_login(int argc, str* argv)
{
  int cr;
  if (!auth_available)
    respond(0,"BAD LOGIN command disabled without SSL/TLS");
  else if (argc != 2)
    respond(0, "BAD LOGIN command requires exactly two arguments");
  else {
    if ((cr = cvm_authenticate_password(cvm, argv[0].s, domain,
					argv[1].s, 1)) == 0)
      do_exec();
    else
      respond(0, "NO LOGIN failed");
    ++auth_count;
    if (auth_max > 0 && auth_count >= auth_max)
      exit(0);
  }
}

void cmd_authenticate(int argc, str* argv)
{
  int i;
  if (argc == 1)
    i = sasl_auth2(&saslauth, argv[0].s, 0);
  else if (argc == 2)
    i = sasl_auth2(&saslauth, argv[0].s, argv[1].s);
  else {
    respond(0, "BAD AUTHENTICATE command requires only one or two arguments");
    return;
  }
  if (i == 0)
    do_exec();
  respond_start(0);
  respond_str("NO AUTHENTICATE failed: ");
  respond_str(sasl_auth_msg(&i));
  respond_end();
    ++auth_count;
    if (auth_max > 0 && auth_count >= auth_max)
      exit(0);
}

void cmd_starttls(void)
{
  if (!tls_available) {
    respond(0, "BAD STARTTLS not availale");
    return;
  }

  respond(0,"OK starting TLS negotiation");
  if (!ucspitls())
    exit(1);

  /* FIX - Reset state? */
  tls_available = 0;
  auth_available = 1;
}

struct command
{
  const char* name;
  void (*fn0)(void);
  void (*fn1)(int argc, str* argv);
};

struct command commands[] = {
  { "CAPABILITY",  cmd_capability, 0 },
  { "NOOP",        cmd_noop,       0 },
  { "LOGOUT",      cmd_logout,     0 },
  { "LOGIN",       0, cmd_login },
  { "AUTHENTICATE", 0, cmd_authenticate },
  { "STARTTLS"   , cmd_starttls,   0 },
  { 0, 0, 0 }
};

static void dispatch_line(void)
{
  struct command* c;
  str_upper(&cmd);
  for (c = commands; c->name != 0; ++c) {
    if (str_diffs(&cmd, c->name) == 0) {
      if (line_argc == 0) {
	if (c->fn0 == 0)
	  respond(0, "BAD Syntax error: command requires arguments");
	else
	  c->fn0();
      }
      else {
	if (c->fn1 == 0)
	  respond(0, "BAD Syntax error: command requires no arguments");
	else
	  c->fn1(line_argc, line_args);
      }
      return;
    }
  }
  respond(0, "BAD Unimplemented command");
}

static int startup(int argc, char* argv[])
{
  const char* tmp;
  if (argc < 2) {
    respond(NOTAG, "NO Usage: imapfront-auth imapd [args ...]");
    return 0;
  }
  if ((tmp = getenv("MAXAUTHFAIL")) != 0)
    auth_max = strtoul(tmp, 0, 10);
  if ((domain = cvm_ucspi_domain()) == 0)
    domain = "unknown";
  nextcmd = argv + 1;
  if ((cvm = getenv("CVM_SASL_PLAIN")) == 0) {
    respond(NOTAG, "NO $CVM_SASL_PLAIN is not set");
    return 0;
  }
  if (!sasl_auth_init(&saslauth)) {
    respond(NOTAG, "NO Could not initialize SASL AUTH");
    return 0;
  }
  
  if ((capability = getenv("IMAP_CAPABILITY")) == 0 &&
      (capability = getenv("CAPABILITY")) == 0)
    capability = "";
  if (strncasecmp(capability, "IMAP4rev1", 9) == 0) capability += 9;
  while (isspace(*capability)) ++capability;
  
  return 1;
}

const int authenticating = 1;
extern void set_timeout(void);

int main(int argc, char* argv[])
{
  if (getenv("UCSPITLS"))
    tls_available = 1;
  if (getenv("AUTH_REQUIRES_TLS"))
    auth_available = 0;
  set_timeout();
  if (!startup(argc, argv)) return 0;
  respond(NOTAG, "OK imapfront ready.");
  while (ibuf_getstr_crlf(&inbuf, &line)) {
    if (parse_line())
      dispatch_line();
  }
  if (ibuf_timedout(&inbuf))
    respond(NOTAG, "NO Connection timed out");
  return 0;
}
